lib/vector-core/buffers/src/disk/mod.rs

```diff
@@ -117,5 +121,11 @@ where
         })?;
 
     let (writer, reader, acker) = leveldb_buffer::Buffer::build(&path, max_size)?;
+    let reader = reader.inspect(|item| {
+        emit(&EventsSent {
+            count: 1,
+            byte_size: size_of_val(item),
```

Comment:
I was initially confused by this too, but `size_of_val` actually just returns the size of the type rather than the size of the value. I believe, instead, you will want to call `allocated_bytes()` which is part of the `ByteSizeOf` trait Brian added in https://github.com/vectordotdev/vector/pull/8396 to allow approximating the size of an event in memory. This will mean adding `ByteSizeOf` as a trait bound on `Item` here.

However, there is another nuance here between memory buffers and disk buffers:

* For memory buffers, I think using the `ByteSizeOf#allocated_bytes` is correct
* For disk buffers, I'm wondering if the metric should be representing the size of the serialized value since that is what is used in the `max_bytes` calculation to ensure the disk buffer doesn't overflow

That would mean, instead, publishing the value around here:

https://github.com/vectordotdev/vector/blob/e10294c7b41b7328f708c03a37740c9e617270d8/lib/vector-core/buffers/src/disk/leveldb_buffer/reader.rs#L158

Where you have access to the serialized bytes size.

@bruceg might have further thoughts though. This distinction between reporting in-memory bytes vs. serialized bytes for events has come up in his work as well.

Reply:
Ah, I see. The explanation and suggestions are great. Will this require moving the `ByteSizeOf` to its own sub-crate (or perhaps moving buffer into vector-core/src)?

Reply:
Hmm, yeah, it does look like it'll require pulling that trait up to be shared. I could see it being its own crate like `internal-event` is now.

Reply:
I think I still prefer to have buffer in its own sub-crate -- its easier to work on this way -- but the benefit is getting closer to outweighed in my mind by the awkwardness we're seeing here. Personally, I'd vote for changing the internal_event crate to something like `core_common` and move `ByteSizeOf` into that.

Reply:
I would not be -1 on moving the buffer code up into `vector_core` itself if others are of a mind to do that.

Reply:
For what it's worth if we start moving in the direction of crate-per-trait I'd see this as a sign we've gone off the rails somewhere.

Reply:
Gotcha, I like the `core_common` idea. Will give that a shot unless there's preference otherwise.